src/int.rs

```diff
@@ -0,0 +1,1409 @@
+                Sign::Positive => I256::MAX,
+                Sign::Negative => I256::MIN,
+            }
+        })
+    }
+
+    /// Wrapping multiplication.
+    pub fn wrapping_mul(self, rhs: Self) -> Self {
+        let (result, _) = self.overflowing_mul(rhs);
+        result
+    }
+
+    /// Returns the sign of `self` to the exponent `exp`.
+    ///
+    /// Note that this method does not actually try to compute the `self` to the
+    /// exponent `exp`, but instead uses the property that a negative number to
+    /// an odd exponent will be negative. This means that the sign of the result
+    /// of exponentiation can be computed even if the actual result is too large
+    /// to fit in 256-bit signed integer.
+    fn pow_sign(self, exp: u32) -> Sign {
+        let is_exp_odd = exp % 2 != 0;
+        if is_exp_odd && self.is_negative() {
+            Sign::Negative
+        } else {
+            Sign::Positive
+        }
+    }
+
+    /// Create 10**n as this type.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the result overflows the type.
+    pub fn exp10(n: usize) -> Self {
+        U256::exp10(n).try_into().expect("overflow")
+    }
+
+    /// Raise self to the power of `exp`.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the result overflows the type in debug mode.
+    pub fn pow(self, exp: u32) -> Self {
+        handle_overflow(self.overflowing_pow(exp))
+    }
+
+    /// Raises self to the power of `exp`.
+    ///
+    /// Returns a tuple of the exponentiation along with a bool indicating
+    /// whether an overflow happened.
+    pub fn overflowing_pow(self, exp: u32) -> (Self, bool) {
+        let sign = self.pow_sign(exp);
+        let (unsigned, overflow_pow) = self.abs_unsigned().overflowing_pow(exp.into());
+        let (result, overflow_conv) = I256::overflowing_from_sign_and_abs(sign, unsigned);
+
+        (result, overflow_pow || overflow_conv)
+    }
+
+    /// Raises self to the power of `exp`. Returns None if overflow occurred.
+    pub fn checked_pow(self, exp: u32) -> Option<Self> {
+        let (result, overflow) = self.overflowing_pow(exp);
+        if overflow {
+            None
+        } else {
+            Some(result)
+        }
+    }
+
+    /// Raises self to the power of `exp`, saturating at the numeric bounds
+    /// instead of overflowing.
+    pub fn saturating_pow(self, exp: u32) -> Self {
+        let (result, overflow) = self.overflowing_pow(exp);
+        if overflow {
+            match self.pow_sign(exp) {
+                Sign::Positive => I256::MAX,
+                Sign::Negative => I256::MIN,
+            }
+        } else {
+            result
+        }
+    }
+
+    /// Wrapping powolute value. Computes `self.pow()`, wrapping around at the
+    /// boundary of the type.
+    pub fn wrapping_pow(self, exp: u32) -> Self {
+        let (result, _) = self.overflowing_pow(exp);
+        result
+    }
```

Comment:
I guess this would be the place where one could implement `{overflowing, checked, saturating, wrapping}` `div` and `mod`. At first glance though, I can't see how division could ever overflow... since division only "blows up" when dividing by real numbers strictly between 0 and 1.

Reply:
`I256::MIN / -1` would overflow, as `-I256::MIN` also overflows. The other subtlety is that `checked_div` returns a `None` when dividing by 0, while the other operations should panic. (This is just to maintain consistency with the signed primitive integers such as `i32`.

Reply:
Awesome, so it does make sense to implement all four versions for each then for consistency even if they are only single cases for some.

Reply:
When in doubt, I just look at what the standard library does with its signed integers: https://doc.rust-lang.org/std/primitive.i32.html

It includes nice documentation and examples that illustrate some of these edge cases.

Reply:
Usually (take a look at the other operations `+ - *`) all of the variants can be implemented in terms of the `overflowing_*` variant of the operation.

Reply:
Yes! thanks for the sick hints.